k2/csrc/ragged_ops.h

```diff
@@ -662,6 +665,64 @@ bool Equal(Ragged<T> &a, Ragged<T> &b) {
   return Equal(a.shape, b.shape) && Equal(a.values, b.values);
 }
 
+
+/*
+  Indexing operation on ragged tensor's shape (indexing axis 0 with
+  a provided array of indexes)
+
+      @param [in] src      Source ragged tensor to index
+      @param [in] indexes  Array of indexes, which will be interpreted
+                           as indexes into axis 0 of `src`,
+                           i.e. with 0 <= indexes[i] < src.Dim0().
+      @param [out]         If non-null, this will be set to a new
+                           Array1<int32_t> containing the indexes
+                           into the elements of an array with shape
+                           'src', that an array with shape 'ans'
+                           would have.  As in:
+                           `ans_values = src_values[*elem_indexes]`.
+
+      @return Returns a ragged shape with
+              `ans.NumAxes() == src.NumAxes()`
+              and `ans.Dim0() == indexes.Dim()`.
+*/
+RaggedShape Index(RaggedShape &src, const Array1<int32_t> &indexes,
+                  Array1<int32_t> *elem_indexes = nullptr);
+
+
+/*
+  Indexing operation on ragged tensor, returns src[indexes], where
+  the elements of `indexes` are interpreted as indexes into the
+  1st axis of `src`.
+
+      @param [in] src      Source ragged tensor to index
+      @param [in] indexes  Array of indexes, which will be interpreted
+                           as indexes into axis 0 of `src`,
+                           i.e. with 0 <= indexes[i] < src.Dim0().
+      @param [out]         If non-null, this will be set to a new
+                           Array1<int32_t> containing the indexes
+                           into src.values that ans.values has,
+                           as in
+                           `ans.values = src.values[value_indexes_out].`
+                           May be useful for backprop.
+
+      @return Returns a ragged tensor with
+              `ans.NumAxes() == src.NumAxes()`
+              and `ans.Dim0() == indexes.Dim()`.
+
+*/
+template <typename T>
+Ragged<T> Index(Ragged<T> &src, const Array1<int32_t> &indexes,
+                Array1<int32_t> *value_indexes_out = nullptr) {
+  Array1<int32_t> value_indexes;
+  RaggedShape ans_shape = Index(src.shape, indexes, &value_indexes);
+  return Ragged<T>(ans_shape,
+                   src.values[value_indexes]);
+  if (value_indexes_out)
+    *value_indexes_out = std::move(value_indexes);
```

Comment:
There is no move assignment operator for `Array1<T>`, so this actually invokes the assignment operator. That is, nothing is moved here. Fixed in #318